defi/insurance/src/lib.rs

```diff
@@ -0,0 +1,250 @@
+            self.policies.insert(ip_resource_def.address(), vault);
+
+            // lock assets
+            let locked = self.assets_pool.take(coverage * supply);
+            self.locked_pool.put(locked)
+        }
+
+        // Purchase specific policy by address
+        pub fn purchase(&mut self, policy_address: Address, insurer: Address, bucket: Bucket) -> Bucket {
+            scrypto_assert!(self.policies.contains_key(&policy_address), "No policy found");
+            scrypto_assert!(bucket.resource_def() == RADIX_TOKEN.into(), "You must purchase policies with Radix (XRD).");
+            
+            // Don't allow an insurer to buy the exact same policy again
+            let purchases = self.purchases.entry(insurer).or_insert(HashMap::new());
+            scrypto_assert!(!purchases.contains_key(&policy_address), "The insurer already has this policy");
+
+            // check if we have free policy in the supply
+            let policies = self.policies.get(&policy_address).unwrap();
+            scrypto_assert!(!policies.is_empty(), "No available policies");
+
+            // take one policy from the supply
+            let policy = policies.take(1);
+            // get metadata
+            let metadata = policy.resource_def().metadata();
+            // check price
+            let price:Decimal = metadata["price"].parse().unwrap();
+            scrypto_assert!(bucket.amount() > price, "Not enough amount to purchase this policy");
+
+            // check duration and setup the end epoch
+            let duration:u64 = metadata["duration"].parse().unwrap();
+            let expires = Context::current_epoch() + duration;
+
+            // get coverage
+            let coverage:Decimal = metadata["coverage"].parse().unwrap();
+
+            // build and mint policy badge with supply equals to coverage
+            let ip_resource_def = ResourceBuilder::new()
+                .metadata("name", "Insurance Purchase Badge")
+                .metadata("symbol", "IPB")
+                .metadata("expires", expires.to_string())
+                .metadata("policy", policy_address.to_string())
+                .new_badge_mutable(self.org_vault.resource_def());
+
+            let ip_badge = self.org_vault.authorize(|badge| {
+                ip_resource_def
+                    .mint(coverage, badge)
+            });
+            
+
+            // burn taken policy
+            self.org_vault.authorize(|badge| {
+                policy.burn(badge);
+            });
+
+            // update the vault
+            let vault = Vault::with_bucket(ip_badge);
+            purchases.insert(policy_address, vault);           
+
+            // take payment
+            let payment = bucket.take(price);
+            self.assets_pool.put(payment);
+
+            // return the rest bucket
+            bucket
+        }
+
+        // Approve payment with specific amount to the insurer using policy address 
+        //#[auth(org_badge)]
```

Comment:
Did you forget to remove the comments on the `auth` macros ?

Reply:
@cbisaillon 
Ah, yes. I have to leave those commented because otherwise  my VSCode highlights everything red with following error (make_policy example):

```
expected 6 arguments, found 7rust-analyzermismatched-arg-count
scrypto_derive
pub macro auth
```

```
expected 6 arguments, found 7rust-analyzermismatched-arg-count
out::Insurance
pub fn make_policy(&self, arg0: String, arg1: Decimal, arg2: Decimal, arg3: u64, arg4: Decimal, auth: ::scrypto::resource::BucketRef)
out::blueprint::Insurance
pub fn make_policy(&mut self, policy_type: String, coverage: Decimal, price: Decimal, duration: u64, supply: Decimal)
```

I wasn't not sure is this something related to my rust-analyzer setup or some exceed settings, so I ended up with commenting auth macros everywhere.